monitor.go

```diff
@@ -0,0 +1,223 @@
+package main
+
+import (
+	"encoding/hex"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/Sirupsen/logrus"
+	raven "github.com/getsentry/raven-go"
+	"github.com/pborman/uuid"
+	"github.com/taskcluster/statsum"
+	tcauth "github.com/taskcluster/taskcluster-client-go/auth"
+)
+
+// NewMonitor creates a new monitor
+func NewMonitor(project string, auth *tcauth.Auth, logLevel string, tags map[string]string) *Monitor {
+	// Create statsumConfigurer
+	statsumConfigurer := func(project string) (statsum.Config, error) {
+		res, err := auth.StatsumToken(project)
+		if err != nil {
+			return statsum.Config{}, err
+		}
+		return statsum.Config{
+			Project: res.Project,
+			BaseURL: res.BaseURL,
+			Token:   res.Token,
+			Expires: time.Time(res.Expires),
+		}, nil
+	}
+
+	// Create logger and parse logLevel
+	logger := logrus.New()
+	switch strings.ToLower(logLevel) {
+	case logrus.DebugLevel.String():
+		logger.Level = logrus.DebugLevel
+	case logrus.InfoLevel.String():
+		logger.Level = logrus.InfoLevel
+	case logrus.WarnLevel.String():
+		logger.Level = logrus.WarnLevel
+	case logrus.ErrorLevel.String():
+		logger.Level = logrus.ErrorLevel
+	case logrus.FatalLevel.String():
+		logger.Level = logrus.FatalLevel
+	case logrus.PanicLevel.String():
+		logger.Level = logrus.PanicLevel
+	default:
+		panic(fmt.Sprintf("Unsupported log-level: %s", logLevel))
+	}
+
+	// Convert tags to logrus.Fields
+	fields := make(logrus.Fields, len(tags))
+	for k, v := range tags {
+		fields[k] = v
+	}
+
+	// Declare monitor so we can reference it in OnError
+	var m *Monitor
+	m = &Monitor{
+		Statsum: statsum.New(project, statsumConfigurer, statsum.Options{
+			OnError: func(err error) { m.ReportWarning(err) },
+		}),
+		Entry: logrus.NewEntry(logger).WithFields(fields),
+		sentry: &sentry{
+			client:  nil,
+			project: project,
+			auth:    auth,
+		},
+	}
+
+	return m
+}
+
+type sentry struct {
+	client     *raven.Client
+	m          sync.Mutex
+	project    string
+	expiration time.Time
+	auth       *tcauth.Auth
+}
+
+func (s *sentry) Client() (*raven.Client, error) {
+	s.m.Lock()
+	defer s.m.Unlock()
+
+	// Refresh sentry DSN if necessary
+	if s.expiration.Before(time.Now()) {
+		// Fetch DSN
+		res, err := s.auth.SentryDSN(s.project)
+		if err != nil {
+			return nil, err
+		}
+		// Create or update DSN for the client
+		if s.client == nil {
+			s.client, err = raven.New(res.Dsn.Secret)
+		} else {
+			err = s.client.SetDSN(res.Dsn.Secret)
+		}
+		if err != nil {
+			return nil, err
+		}
+		// Set expiration, so we remember to refresh
+		s.expiration = time.Time(res.Expires)
+	}
+
+	return s.client, nil
+}
+
+type Monitor struct {
+	*statsum.Statsum
+	*logrus.Entry
+	*sentry
+	tags   map[string]string
+	prefix string
+}
+
+func (m *Monitor) Time(name string, fn func()) {
```

Comment:
Doc comment would be useful here (I know it is simple, but I've no idea what the statsum Time function does).